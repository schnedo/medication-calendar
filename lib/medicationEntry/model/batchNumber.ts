export default interface BatchNumber {
  readonly number: number;
}

const prefix = "Ch.-B.";
export function format(batchNumber: BatchNumber): string {
  return `${prefix} ${batchNumber.number}`;
}
