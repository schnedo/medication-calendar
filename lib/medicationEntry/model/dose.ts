export default interface Dose {
  readonly amount: number;
}

const unit = "ml";

export function format(dose: Dose): string {
  return `${dose.amount} ${unit}`;
}
