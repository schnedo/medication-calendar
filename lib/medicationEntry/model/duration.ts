export default interface Duration {
  minutes: number;
  hours: number;
}

function formatPart(number: number | undefined) {
  if (number === undefined) {
    return "00";
  }
  const raw = number.toString();
  return raw.length === 1 ? `0${raw}` : raw;
}

export function format(duration: Duration): string {
  return `${formatPart(duration.hours)}:${formatPart(duration.minutes)}`;
}
