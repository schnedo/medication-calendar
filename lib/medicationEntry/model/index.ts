export { default as BatchNumber } from "./batchNumber";
export { default as Dose } from "./dose";
export { default as Duration } from "./Duration";
export { default as Medicament } from "./medicament";
export { default as Medication } from "./medication";
export { default as MedicationEntry } from "./medicationEntry";

export * as Repository from "./repository";
