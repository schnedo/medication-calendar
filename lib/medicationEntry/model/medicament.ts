import BatchNumber from "./batchNumber";

export default interface Medicament {
  readonly name: string;
  readonly batchNumber: BatchNumber;
}
