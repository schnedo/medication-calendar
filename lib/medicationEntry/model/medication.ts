import Dose from "./dose";
import Medicament from "./medicament";
import { v4 } from "uuid";

export default class Medication {
  readonly id: string = v4();

  constructor(readonly dose: Dose, readonly medicament: Medicament) {}
}
