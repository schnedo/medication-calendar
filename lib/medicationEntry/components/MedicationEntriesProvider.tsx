import {
  createContext,
  ReactElement,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from "react";
import { MedicationEntry } from "../model";
import { Entity, Repository } from "../../storage";

const repository = new Repository<MedicationEntry>({
  storeName: "medicationEntries",
});

export interface MedicationEntriesContext {
  entries: (MedicationEntry & Entity)[] | null;
  saveEntry: (entry: MedicationEntry) => Promise<void>;
  deleteEntry: (entry: MedicationEntry) => Promise<void>;
}

const notInitialized = () =>
  Promise.reject(new Error("MedicationEntriesProvider not initialized yet"));

const defaultContext: MedicationEntriesContext = {
  entries: null,
  saveEntry: notInitialized,
  deleteEntry: notInitialized,
};

const medicationEntriesContext = createContext(defaultContext);
medicationEntriesContext.displayName = "MedicationEntriesContext";

export const useMedicationEntries = (): MedicationEntriesContext =>
  useContext(medicationEntriesContext);

export interface MedicationEntriesProviderProps {
  children?: ReactNode;
}

export default function MedicationEntriesProvider({
  children,
}: MedicationEntriesProviderProps): ReactElement {
  const [value, setValue] = useState<MedicationEntriesContext>(defaultContext);
  useEffect(() => {
    repository.getAll().then((entries) => {
      setValue({
        entries,
        saveEntry: async (entry) => {
          const withId = await repository.save(entry);
          setValue((old) => ({
            ...old,
            entries: [...(old.entries ?? []), withId],
          }));
        },
        deleteEntry: async (entry) => {
          await repository.remove(entry);
          setValue((old) => ({
            ...old,
            entries: old.entries?.filter(({ id }) => id !== entry.id) ?? [],
          }));
        },
      });
    });
  }, []);
  return (
    <medicationEntriesContext.Provider value={value}>
      {children}
    </medicationEntriesContext.Provider>
  );
}
