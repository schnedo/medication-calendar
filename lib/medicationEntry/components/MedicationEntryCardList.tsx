import { ReactElement } from "react";
import { MedicationEntryCard } from "./index";
import { Box, CircularProgress, makeStyles } from "@material-ui/core";
import { MedicationEntry } from "../model";

const useStyle = makeStyles((theme) => ({
  container: {
    "&>*": {
      marginBottom: theme.spacing(1),
      "&:last-child": {
        marginBottom: 0,
      },
    },
  },
}));

export interface MedicationEntryCardListSelection {
  medicationEntry: MedicationEntry;
  pointerPosition: {
    clientX: number;
    clientY: number;
  };
}

export interface MedicationEntryCardListProps {
  medicationEntries: MedicationEntry[] | null;
  onSelect: (
    medicationEntryCardListSelection: MedicationEntryCardListSelection,
  ) => void;
}

export default function MedicationEntryCardList({
  medicationEntries,
  onSelect,
}: MedicationEntryCardListProps): ReactElement {
  const { container } = useStyle();
  return medicationEntries === null ? (
    <CircularProgress />
  ) : (
    <Box className={container}>
      {medicationEntries.map((entry) => (
        <MedicationEntryCard
          key={entry.id}
          medicationEntry={entry}
          onClick={({ clientX, clientY }) => {
            onSelect({
              medicationEntry: entry,
              pointerPosition: { clientX, clientY },
            });
          }}
        />
      ))}
    </Box>
  );
}
