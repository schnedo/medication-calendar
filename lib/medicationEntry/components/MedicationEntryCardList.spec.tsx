import { render } from "@testing-library/react";
import MedicationEntryCardList from "./MedicationEntryCardList";
import { mock } from "@userlike/joke";
import { MedicationEntry } from "../model";
import { MedicationEntriesContext } from "./MedicationEntriesProvider";

const { useMedicationEntries } = mock(import("./MedicationEntriesProvider"));
const { default: MedicationEntryCard } = mock(import("./MedicationEntryCard"));

function getMockContext(
  entries: MedicationEntry[] | null,
): MedicationEntriesContext {
  return {
    entries,
    saveEntry: jest.fn(),
    deleteEntry: jest.fn(),
  };
}

describe.skip("MedicationEntryCardList", () => {
  beforeAll(() => {
    MedicationEntryCard.mockReturnValue(<div>mocked MedicationEntryCard</div>);
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render medication card list", async () => {
    expect.hasAssertions();

    const entries: MedicationEntry[] = [
      {
        id: "1",
        date: new Date(2020, 1, 10),
        medications: [],
        duration: { minutes: 15, hours: 1 },
        comments: "",
        bodyMass: {
          amount: 70,
        },
      },
    ];
    useMedicationEntries.mockReturnValue(getMockContext(entries));

    const { container } = render(<MedicationEntryCardList />);

    expect(container).toMatchSnapshot();
  });

  it("should render loading animation when entries are null", async () => {
    expect.hasAssertions();

    useMedicationEntries.mockReturnValue(getMockContext(null));

    const { container } = render(<MedicationEntryCardList />);

    expect(container).toMatchSnapshot();
  });
});
