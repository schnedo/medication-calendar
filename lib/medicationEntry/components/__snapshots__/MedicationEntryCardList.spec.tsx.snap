// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MedicationEntryCardList should render loading animation when entries are null 1`] = `
<div>
  <div
    class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;

exports[`MedicationEntryCardList should render medication card list 1`] = `
<div>
  <ul
    class="MuiList-root MuiList-padding"
  >
    <li
      class="MuiListSubheader-root MuiListSubheader-sticky MuiListSubheader-gutters"
    >
      February 2020
    </li>
    <li
      class="MuiListItem-root MuiListItem-gutters"
    >
      <div>
        mocked MedicationEntryCard
      </div>
    </li>
  </ul>
</div>
`;
